File: trainingregional2018/contest1/P3.py
import sys

def get_cases():
	casos = []
	ccaso = []
	for line in sys.stdin:
		line = line.strip()
		if line == "":
			if len(ccaso) > 0:
				casos.append(ccaso)
				ccaso = []
		else:
			ccaso.append(line)
	if len(ccaso) > 0:
		casos.append(ccaso)
	return casos

def solve_case(caso):
	digits = caso[0]
	mp = {}
	for i, c in enumerate(digits):
		mp[c] = i
	
	ans = []
	for i in range(1, len(caso)):
		suma = 0
		a = 0
		b = len(digits)
		for c in caso[i]:
			a = max(a, mp[c])
		a += 1

		pot = [1]*(b - a + 1)
		for j, c in enumerate(reversed(caso[i])):
			suma += mp[c]*sum(pot)
			for k in range(a, b + 1):
				pot[k-a] *= k
		ans.append(suma)
	return ans

casos = get_cases()
pri = True
for caso in casos:
	if not pri:
		print("")
	pri = False
	for x in solve_case(caso):
		print(x)

File: trainingregional2018/contest1/P3.cpp
#include <bits/stdc++.h>

#define REP(i, n) for(int i = 0; i < n; i++)
#define REPR(i, a, b) for (int i = a; i < b; i++)
#define clr(t, val) memset(t, val, sizeof(t))
#define all(v) v.begin(), v.end()
#define SZ(v) ((int)(v).size())
#define TEST(x) cerr << "test " << #x << " " << x << endl;

using namespace std;

clock_t _inicio = clock();

typedef long long Long;
typedef vector <int> vInt;
typedef pair <int, int> Pair;


vector <vector <string> > getCases() {
    vector <vector <string> > ans;
    string line;
    vector <string> currentCase;
    while (getline(cin, line)) {
        int i = 0;
        REP(j, SZ(line)) 
            if (line[j] != ' ') line[i++] = line[j];
        while (SZ(line) > i) line.pop_back();

        if (line == "") {
            if (SZ(currentCase)) ans.push_back(currentCase);
            currentCase.clear();
        } else {
            currentCase.push_back(line);
        }
    }

    if (SZ(currentCase)) ans.push_back(currentCase);
    return ans;
}

vector <Long> solveCase(vector<string> &caso) {
    map <char, int> mp;
    REP(i, SZ(caso[0])) {
        mp[caso[0][i]] = i;
    }
    vector <Long> ans;

    REPR(i, 1, SZ(caso)) {
        string num = caso[i];
        int a = 0, b = SZ(caso[0]);
        REP(j, SZ(num)) {
            a = max(a, mp[num[j]]);
        }
        a++;

        reverse(all(num));
        vector <Long> pot(b - a + 1, 1);
        Long sum = 0;
        REP(j, SZ(num)) {
            Long lsum = 0;
            REP(k, SZ(pot)) lsum += pot[k];
            sum += mp[num[j]]*lsum;
            REPR(k, a, b + 1) {
                pot[k-a] *= k;
            }
        }
        ans.push_back(sum);
    }

    return ans;
}

int main(int nargs, char **args) {
    vector <vector <string> > casos = getCases();

    bool pri = true;
    for (vector <string> &caso: casos) {
        vector <Long> sol = solveCase(caso);
        if (!pri) {
            cout << endl;
        } else {
            pri = false;
        }

        for (Long x: sol) {
            cout << x << endl;
        }
    }
    // cerr << "Time elapsed: " << (clock() - _inicio)/1000 << " ms" << endl;
    return 0;
}
